Various audit fixes

yondonfu opened the pull request:
This PR implements a few fixes related to the external security audit. 

Summary of changes:

- Make `getTicketHash()` public so that off-chain hashing implementations can be more easily checked against the output from the EVM
- Add a gas report test for redeeming winning tickets
- Use Node v10.17.0 and upgrade a few dependencies
- Add some notes in the README about ABIEncoderV2 usage and ERC20 functions
- Add comments about calling setters for upgradeable contracts as well as shadowed functions in ManagerProxy
- Add an integration test demonstrating that honest transcoders are protected against ticket front running scenarios as long as they don't hold on to more value than their reserve allocation

See the commit log for more details.

kyriediculous commented:
I assume 6aebc09 - `use node v10.7.0` is actually using `v10.17.0` as that is the latest release for v10.x. ? Could we fix up the commit message if that's the case?

yondonfu commented:
> Oh it is 10.7.0 , any particular reason for this release instead of the latest v10.17.0

For some reason I thought that there was a dependency problem with v10.17.0, but I just tried again and there does not seem to be any problems...Rebased and replaced the commit with one that uses Node v10.17.0

yondonfu commented on README.md:
```diff
@@ -24,8 +24,26 @@ The contracts are based off of the [technical protocol specification](https://gi
+
 ### Install 
 
+Make sure Node v10.7.0 installed.
```
Fixed in 9eb1013

yondonfu commented:
Rebased

kyriediculous commented:
Good to go once checks pass ! 🏁
